### ISSUE TITLE
responsiveness of "Reset All" and "Start Over" buttons

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/natural-selection#61]
Related to https://github.com/phetsims/tandem/issues/173 (PhetioGroup is not scalable).

As the number of bunnies increases, we start to experience problems with responsiveness and performance. This is even before starting to collect the large amounts of data mentioned in #46.

When bunnies have taken over the world, a dialog is displayed.  After closing that dialog, a "Start Over" button is displayed.  Pressing that "Start Over" button (or the "Reset All" button) results in a significant lag, as all of the bunnies in the PhetioGroup are disposed one at a time.

Is this lag acceptable?

To reproduce:
1. Run https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.31/phet/natural-selection_en_phet.html?secondsPerGeneration=1
2. Go to the _Lab_ screen
3. Press the "Add a Mate" button
4. Wait for "Bunnies have taken over the world dialog" to display.
5. Close the dialog.
6. Press "Start Over" button.  Observer the lag time (almost 2 seconds on my MacBook Pro).

The above example is with mostly live bunnies.  When the sim implementation is more fleshed out and we many dead bunnies, the responsiveness will be worse.

**Comment by pixelzoom** [comment]
I noticed that this pause is significantly shorter in the built version.  So there may be some expensive assertions that are responsible for some of the delay.

**Comment by pixelzoom** [comment]
Running unbuilt on a MacBookPro16,1, in Chrome. 

With an initial population of 1, and running until bunnies take over the world (1434 Bunny instances):

With `?ea`:
Pressing "Start Over" calls `model.reset`, and takes 3722 ms. 
Of that, 3716 ms is spend in `bunnyGroup.clear`.

Without `?ea`:
Pressing "Start Over" calls `model.reset`, and takes 101 ms. 
Of that, 97 ms is spend in `bunnyGroup.clear`.

`clear` results in multiple calls to `elementDisposedEmitter`. `bunnyGroup.elementDisposedEmitter` has 1 listener that looks like this:

```js
    // When a bunny is disposed...
    this.bunnyGroup.elementDisposedEmitter.addListener( bunny => {
      assert && assert( bunny instanceof Bunny, 'invalid bunny' );
      this.liveBunnies.contains( bunny ) && this.liveBunnies.remove( bunny );
      this.deadBunnies.contains( bunny ) && this.deadBunnies.remove( bunny );
      this.totalNumberOfBunniesProperty.value = this.bunnyGroup.length;
 } );
```

If I comment out that listener, `bunnyGroup.clear` still takes 3650 ms with `?ea` and 96 ms without `?ea`.   So I think that confirms that PhetioGroup.clear is taking a lot of time for some reason, even with no listeners to `elementDisposedEmitter`.   And there must be some performance-intensive assertion involved with `?ea`.   

Thoughts from the PhET-iO team?

**Comment by samreid** [comment]
The profiler suggests 88% of this time is being spent in dispose:

![image](https://user-images.githubusercontent.com/679486/80662632-fb837500-8a4e-11ea-9970-1ed24cf4a2e2.png)

**Comment by pixelzoom** [comment]
Bunny `dispose` isn't doing a whole lot.  It's disposing of 2 Properties, 1 Emitter, 1 Genotype (a PhetioObject), 3 GenePairs (PhetioObjects) and 1 Phenotype (PhetioObject).  Those additional PhetioObjects have nothing subclass-specific that is disposed.   So I don't see anything obvious in sim code that can be optimized.

@samreid Would you some time to pair on this with me, to drill down and identify what part of `dispose` is so costly with `?ea` ?

**Comment by samreid** [comment]
I tired profiling without ?ea and observed nearly 400ms (with profiler on). In this case, it 50% of the time is in Scenery.updateDisplay and more like 30% in dispose.

> Bunny dispose isn't doing a whole lot.

This likely cascades to the BunnyNode dispose.  Also, I noticed some time being spent in accessibility dispose:

![image](https://user-images.githubusercontent.com/679486/80668353-ea426480-8a5e-11ea-8160-7f1e917399b2.png)

Maybe we will request an optimized version of setAriaLabelledbyAssociations for use in disposal?  Tagging @jessegreenberg in case he has ideas about that part.

I'm happy to schedule a meeting to take a closer look.  Let's schedule over slack.

**Comment by pixelzoom** [comment]
> This likely cascades to the BunnyNode dispose. 

BunnyNode `dispose` unlinks 1 listener and disposes 1 Property.   So I doubt that BunnyNode is responsible for the increased delay with `?ea`.  More likely it's Node or PhetioObject.

**Comment by samreid** [comment]
BunnyNode dispose takes up around 27%, but that is due to Node code, not code specific to BunnyNode.  When assertions are enabled, there is a `setTimeout` in PhetioObject.dispose which shows it is taking 63% of the time during this step.

![image](https://user-images.githubusercontent.com/679486/80736407-52c92a00-8acf-11ea-8ade-a5fe9d6fae2b.png)

**Comment by pixelzoom** [comment]
As @samreid mentioned in the previous comment, PhetioObject `dispose` contains this assertion:

```js
    // In order to support the structured data stream, PhetioObjects must end the messages in the correct
    // sequence, without being interrupted by dispose() calls.  Therefore, we do not clear out any of the state
    // related to the endEvent.  Note this means it is acceptable (and expected) for endEvent() to be called on
    // disposed PhetioObjects.
    //
    // The phetioEvent stack should resolve by the next clock tick, so that's when we check it.
    assert && setTimeout( () => { // eslint-disable-line bad-sim-text
      assert && assert( self.phetioMessageStack.length === 0, 'phetioMessageStack should be clear' );
    }, 0 );
```

Reading the comment, I'm not clear on the purpose of this assertion, and why it's not a problem if it occurs when assertion are disabled.   

Commenting out that assertion results in acceptable performance of "Start Over" and "Reset" in Natural Selection. 

@samreid let's discuss via Zoom.

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/natural-selection/issues/61#issuecomment-621588910, @samreid said:
> Maybe we will request an optimized version of `setAriaLabelledbyAssociations` for use in disposal? Tagging @jessegreenberg in case he has ideas about that part.

I'd also like to hear from @jessegreenberg.  `Node.dispose` calls `ParallelDOM.disposeAccessibility`.   Has this method, including `setAriaLabelledbyAssociations`, been optimized?

**Comment by pixelzoom** [comment]
@samreid and I investigated the performance of `PhetioObject.dispose` via Zoom.  Changing `setTimeout` to `phet.axon.setTimeout` resulted in a 2x performance increase.  But adding `runOnNextFrame` method to Sim.js resulted in a 12x improvement, so we went with that approach, which also provides clarity to what is actually going on.  Sim.js changes are in the above commits.

**Comment by pixelzoom** [comment]
Labeling for design meeting, to get input from team on what acceptable responsiveness is.

**Comment by pixelzoom** [comment]
Up to 1 second response is acceptable.

**Comment by pixelzoom** [comment]
Actually, I'm going to leave this open and deferred, so we can revisit later.

**Comment by pixelzoom** [comment]
After implementing all graphs...  These buttons feel a little sluggish in the unbuilt version with assertions enabled.  But in the the built version, they feel plenty responsive.  So I'm going to close this issue.
